csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,210 @@
+                    // The connection cannot be trust clear the connection and try again.
+                    connection = null;
+                }
+            }
+            while (connection == null);
+            return connection;
+
+            async Task<Connection> PerformConnectAsync(
+                Endpoint endpoint,
+                NonSecure preferNonSecure,
+                object? label)
+            {
+                try
+                {
+                    Debug.Assert(ConnectTimeout > TimeSpan.Zero);
+                    using var source = new CancellationTokenSource(ConnectTimeout);
+                    CancellationToken cancel = source.Token;
+                    Connection connection = await endpoint.ConnectAsync(preferNonSecure,
+                                                                        label,
+                                                                        cancel).ConfigureAwait(false);
+                    lock (_mutex)
+                    {
+                        if (!_outgoingConnections.TryGetValue((endpoint, label), out List<Connection>? list))
+                        {
+                            list = new List<Connection>();
+                            _outgoingConnections[(endpoint, label)] = list;
+                        }
+
+                        list.Add(connection);
+                    }
+                    // Set the callback used to remove the connection from the factory.
+                    connection.Remove = connection => Remove(connection, label);
+                    return connection;
+                }
+                catch (TransportException)
+                {
+                    _transportFailures[endpoint] = DateTime.Now;
+                    throw;
+                }
+                finally
+                {
+                    lock (_mutex)
+                    {
+                        _pendingOutgoingConnections.Remove((endpoint, label));
+                    }
+                }
+            }
+        }
+
+        internal Connection? GetConnection(List<Endpoint> endpoints, NonSecure preferNonSecure, object? label)
+        {
+            lock (_mutex)
+            {
+                foreach (Endpoint endpoint in endpoints)
+                {
+                    if (_outgoingConnections.TryGetValue((endpoint, label),
+                                                         out List<Connection>? connections) &&
+                        connections.FirstOrDefault(
+                            connection => connection.IsActive && connection.CanTrust(preferNonSecure))
+                        is Connection connection)
+                    {
+                        return connection;
+                    }
+                }
+                return null;
+            }
+        }
+
+        internal IEnumerable<Endpoint> OrderEndpointsByTransportFailures(IEnumerable<Endpoint> endpoints)
+        {
+            // Purge expired transport failures
```

Comment:
I would add a TODO to rework this code in a follow-up PR.

I think "purging" the _transportFailures for each single lookup is too much. I would also optimize the common situation where there is a single endpoint.

Reply:
Added a TODO